src/spawn.c

```diff
@@ -0,0 +1,1121 @@
+}
+
+
+static void spawn_capture_text(const gchar *text, gpointer user_data)
+{
+	g_string_append((GString *) user_data, text);
+}
+
+
+typedef struct _SpawnCapture
+{
+	gchar **stdout_text;
+	GString *stdout_str;
+	gchar **stderr_text;
+	GString *stderr_str;
+	gint *exit_status;
+} SpawnCapture;
+
+
+static void spawn_capture_exit(gint status, gpointer user_data)
+{
+	SpawnCapture *sp = (SpawnCapture *) user_data;
+
+	if (sp->stdout_text)
+		*sp->stdout_text = g_string_free(sp->stdout_str, FALSE);
+
+	if (sp->stderr_text)
+		*sp->stderr_text = g_string_free(sp->stderr_str, FALSE);
+
+	if (sp->exit_status)
+		*sp->exit_status = status;
+
+	g_free(user_data);
+}
+
+
+/**
+ *  Executes a child program and captures it's output.
+ *
+ *  Either a command line or an argument vector must be passed. If both are present, the
+ *  argument vector is appended to the command line. An empty command line is not allowed.
+ *
+ *  If a @a block_window is specified (recommended), the function will not return until the
+ *  child exits, but all other windows (if any) will receive events. Otherwise it'll return,
+ *  and it's your responsibility to keep @a stdout_text, @a stderr_text and @a exit_status
+ *  valid until the execution finishes.
+ *
+ *  @a stdin_text will be gfree()-d automatically when it's sent entirely to the child, or
+ *  when the child exits.
+ *
+ *  @a stdout_text and @a stderr_text will be set when the child exits.
+ *
+ *  @param working_directory child's current working directory, or @c NULL.
+ *  @param command_line child program and arguments, or @c NULL.
+ *  @param argv child's argument vector, or @c NULL.
+ *  @param envp child's environment, or @c NULL.
+ *  @param block_window window whose message loop to block until the child exits, or @c NULL.
+ *  @param stdin_text text to send to childs's stdin, or @c NULL.
+ *  @param stdout_text return location for the child's stdout, or NULL.
+ *  @param stderr_text return location for the child's stderr, or NULL.
+ *  @param exit_status return location for the child exit code, or NULL.
+ *  @param error return location for error.
+ *
+ *  @return @c TRUE on success, @c FALSE if an error was set.
+ **/
+gboolean spawn_with_capture(const gchar *working_directory, const gchar *command_line,
+	gchar **argv, gchar **envp, GtkWindow *block_window, gchar *stdin_text,
+	gchar **stdout_text, gchar **stderr_text, gint *exit_status, GError **error)
+{
+	SpawnCapture *sp = g_new(SpawnCapture, 1);
+
+	sp->stdout_text = stdout_text;
+	sp->stdout_str = stdout_text ? g_string_sized_new(1024) : NULL;
+	sp->stderr_text = stderr_text;
+	sp->stderr_str = stderr_text ? g_string_sized_new(1024) : NULL;
+	sp->exit_status = exit_status;
+
+	return spawn_with_callbacks(working_directory, command_line, argv, envp, block_window,
+		stdin_text, stdout_text ? spawn_capture_text : NULL, sp->stdout_str,
+		stderr_text ? spawn_capture_text : NULL, sp->stderr_str, spawn_capture_exit, sp,
+		NULL, error);
+}

```

Comment:
"If a @a block_window is specified, the function will not return until the child exits, but all other windows (if any) will receive events."

block_window = NULL -> async execution.

block_window != NULL -> sync and block the specified window; pass Geany main window for sync execution under Geany. Unlike g_spawn_sync, you can still pass callbacks and receive the child's message in real time if you want, instead of waiting for a big gchar *standard_output and *standard_error at the end. You can pass stdin_text to send too.